File: app/python/ml_tdd/src/ml/survived_classifier.py
"""Survived 生存予測器モジュール."""

import os
import pickle
from typing import Optional, Tuple

import pandas as pd
from sklearn import tree


class SurvivedClassifier:
    """客船沈没事故の生存を予測する分類モデル.

    Attributes:
        max_depth: 決定木の最大深度
        class_weight: クラスの重み付け（'balanced' で自動調整）
        model: 訓練済みの決定木モデル（未訓練時は None）
    """

    def __init__(self, max_depth: int = 9, class_weight: str = "balanced") -> None:
        """初期化.

        Args:
            max_depth: 決定木の最大深度（デフォルト: 9）
            class_weight: クラス重み付け方法（デフォルト: 'balanced'）

        Raises:
            ValueError: max_depth が 1 未満の場合
        """
        if max_depth < 1:
            raise ValueError("max_depth must be at least 1")

        self.max_depth = max_depth
        self.class_weight = class_weight
        self.model: Optional[tree.DecisionTreeClassifier] = None

    def load_data(
        self, file_path: str, preprocess: bool = True
    ) -> Tuple[pd.DataFrame, pd.Series]:
        """CSV ファイルからデータを読み込む.

        Args:
            file_path: CSV ファイルのパス
            preprocess: 前処理を実行するかどうか（デフォルト: True）

        Returns:
            特徴量 DataFrame と目的変数 Series のタプル

        Raises:
            FileNotFoundError: ファイルが存在しない場合
            ValueError: 必要な列が不足している場合
        """
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")

        # データの読み込み
        df = pd.read_csv(file_path)

        # 必要な列の存在確認
        required_columns = [
            "Pclass",
            "Age",
            "SibSp",
            "Parch",
            "Fare",
            "Sex",
            "Survived",
        ]
        missing_columns = set(required_columns) - set(df.columns)
        if missing_columns:
            raise ValueError(f"Missing columns: {missing_columns}")

        # 前処理の実行（オプション）
        if preprocess:
            df = self._preprocess_data(df)

        # 特徴量と目的変数の分割
        feature_columns = ["Pclass", "Age", "SibSp", "Parch", "Fare", "male"]
        if not preprocess:
            # 前処理なしの場合、Sex 列をそのまま使う（テスト用）
            feature_columns = ["Pclass", "Age", "SibSp", "Parch", "Fare", "Sex"]

        X = df[feature_columns]
        y = df["Survived"]

        return X, y

    def _preprocess_data(self, df: pd.DataFrame) -> pd.DataFrame:
        """データの前処理（内部メソッド）.

        Args:
            df: 元の DataFrame

        Returns:
            前処理済みの DataFrame
        """
        df = self._preprocess_age(df)
        df = self._encode_sex(df)
        return df

    def _preprocess_age(self, df: pd.DataFrame) -> pd.DataFrame:
        """Age 列の欠損値をグループ別中央値で補完.

        Pclass と Survived のグループごとに中央値を計算し、欠損値を補完します。

        Args:
            df: 元の DataFrame

        Returns:
            Age が補完された DataFrame
        """
        df_copy = df.copy()

        # グループ別の中央値を計算
        group_medians = df_copy.groupby(["Pclass", "Survived"])["Age"].median()

        # 各行の欠損値を対応するグループの中央値で補完
        for idx, row in df_copy[df_copy["Age"].isnull()].iterrows():
            pclass = row["Pclass"]
            survived = row["Survived"]
            median_age = group_medians.get((pclass, survived))

            if pd.notna(median_age):
                df_copy.loc[idx, "Age"] = median_age
            else:
                # グループに該当がない場合は全体の中央値を使用
                df_copy.loc[idx, "Age"] = df_copy["Age"].median()

        return df_copy

    def _encode_sex(self, df: pd.DataFrame) -> pd.DataFrame:
        """Sex 列をダミー変数に変換.

        Args:
            df: 元の DataFrame

        Returns:
            Sex がダミー変数化された DataFrame
        """
        df_copy = df.copy()

        # Sex をダミー変数化（drop_first=True で female を削除）
        sex_dummies = pd.get_dummies(df_copy["Sex"], drop_first=True, dtype=int)
        df_encoded = pd.concat([df_copy.drop("Sex", axis=1), sex_dummies], axis=1)

        return df_encoded

    def train(self, X_train: pd.DataFrame, y_train: pd.Series) -> None:
        """モデルを訓練する.

        Args:
            X_train: 訓練用特徴量
            y_train: 訓練用目的変数
        """
        self.model = tree.DecisionTreeClassifier(
            max_depth=self.max_depth, class_weight=self.class_weight, random_state=0
        )
        self.model.fit(X_train, y_train)

    def predict(self, X_test: pd.DataFrame) -> pd.Series:
        """予測を行う.

        Args:
            X_test: テスト用特徴量

        Returns:
            予測結果

        Raises:
            RuntimeError: モデルが訓練されていない場合
        """
        if self.model is None:
            raise RuntimeError("Model has not been trained yet. Call train() first.")

        return self.model.predict(X_test)

    def evaluate(self, X_test: pd.DataFrame, y_test: pd.Series) -> float:
        """モデルを評価する.

        Args:
            X_test: テスト用特徴量
            y_test: テスト用目的変数

        Returns:
            正解率（accuracy）

        Raises:
            RuntimeError: モデルが訓練されていない場合
        """
        if self.model is None:
            raise RuntimeError("Model has not been trained yet. Call train() first.")

        score: float = self.model.score(X_test, y_test)
        return score

    def save_model(self, file_path: str) -> None:
        """訓練済みモデルをファイルに保存する.

        Args:
            file_path: 保存先ファイルパス

        Raises:
            RuntimeError: モデルが訓練されていない場合
        """
        if self.model is None:
            raise RuntimeError("Model has not been trained yet. Call train() first.")

        with open(file_path, "wb") as f:
            pickle.dump(self.model, f)

    def load_model(self, file_path: str) -> None:
        """保存されたモデルをファイルから読み込む.

        Args:
            file_path: 読み込むファイルパス

        Raises:
            FileNotFoundError: ファイルが存在しない場合
        """
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")

        with open(file_path, "rb") as f:
            self.model = pickle.load(f)

File: app/typescript/src/models/SurvivedClassifier.ts
import { DataFrame, IDataFrame } from 'data-forge';
import * as fs from 'fs';
import { DecisionTreeClassifier } from 'ml-cart';

interface SurvivedClassifierOptions {
  maxDepth?: number;
}

interface AgeMapping {
  [key: string]: number;
}

/**
 * Survived 生存予測モデル
 * タイタニック号の乗客データから生存者を予測する分類モデル
 */
export class SurvivedClassifier {
  public readonly maxDepth: number;
  private model: DecisionTreeClassifier | null = null;
  private data?: DataFrame;
  private X_train?: number[][];
  private y_train?: number[];
  private X_test?: number[][];
  private y_test?: number[];
  private featureNames: string[] = ['Pclass', 'Age', 'SibSp', 'Parch', 'Fare', 'male'];

  constructor(options: SurvivedClassifierOptions = {}) {
    this.maxDepth = options.maxDepth ?? 9;

    if (this.maxDepth < 1) {
      throw new Error('max_depth must be at least 1');
    }
  }

  isTrained(): boolean {
    return this.model !== null;
  }

  /**
   * CSVファイルからデータを読み込む
   */
  async loadData(
    filePath: string,
    preprocess: boolean = true
  ): Promise<{ X: IDataFrame; y: IDataFrame }> {
    // ファイルの存在確認
    try {
      await fs.promises.access(filePath);
    } catch {
      throw new Error(`File not found: ${filePath}`);
    }

    // データの読み込み
    const fileContent = await fs.promises.readFile(filePath, 'utf-8');

    // CSV を手動でパース（CRLF と LF の両方に対応）
    const lines = fileContent.trim().split(/\r?\n/);
    const headers = lines[0].split(',').map((h) => h.trim());

    const rows = lines.slice(1).map((line) => {
      const values = line.split(',').map((v) => v.trim());
      const obj: any = {};
      headers.forEach((header, index) => {
        const value = values[index];
        // 空文字列を null に変換
        obj[header] = value === '' ? null : value;
      });
      return obj;
    });

    let df = new DataFrame(rows);

    // 必要な列の存在確認
    const requiredColumns = [
      'Pclass',
      'Age',
      'SibSp',
      'Parch',
      'Fare',
      'Sex',
      'Survived',
    ];
    const actualColumns = df.getColumnNames();
    const missingColumns = requiredColumns.filter(
      (col) => !actualColumns.includes(col)
    );

    if (missingColumns.length > 0) {
      throw new Error(`Missing columns: ${missingColumns.join(', ')}`);
    }

    // 前処理の実行（オプション）
    if (preprocess) {
      const preprocessed = this.preprocessData(df);
      df = new DataFrame(preprocessed.toArray());
    }

    // 特徴量と目的変数の分割
    let featureColumns = ['Pclass', 'Age', 'SibSp', 'Parch', 'Fare', 'male'];
    if (!preprocess) {
      featureColumns = ['Pclass', 'Age', 'SibSp', 'Parch', 'Fare', 'Sex'];
    }

    const X = df.subset(featureColumns);
    const y = df.getSeries('Survived');

    // データを保存
    this.data = df;

    return { X: X as any, y: y as any };
  }

  /**
   * データを訓練用とテスト用に分割する
   */
  splitData(testSize: number = 0.2): void {
    if (!this.data) {
      throw new Error('Data not loaded. Call loadData() first.');
    }

    // データを配列に変換してシャッフル
    const dataArray = this.data.toArray();
    const shuffled = this.shuffleArray(dataArray);

    // 分割ポイントを計算
    const totalSize = shuffled.length;
    const trainSize = Math.floor(totalSize * (1 - testSize));

    // 訓練データとテストデータに分割
    const trainArray = shuffled.slice(0, trainSize);
    const testArray = shuffled.slice(trainSize);

    const trainData = new DataFrame(trainArray);
    const testData = new DataFrame(testArray);

    // 特徴量とラベルを抽出
    this.X_train = this.extractFeatures(trainData);
    this.y_train = this.extractLabels(trainData);
    this.X_test = this.extractFeatures(testData);
    this.y_test = this.extractLabels(testData);
  }

  /**
   * 配列をシャッフルする（Fisher-Yates アルゴリズム）
   */
  private shuffleArray<T>(array: T[]): T[] {
    const shuffled = [...array];
    for (let i = shuffled.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
    }
    return shuffled;
  }

  /**
   * 特徴量を抽出する
   */
  private extractFeatures(data: DataFrame): number[][] {
    const rows = data.toArray();
    return rows.map((row: any) => this.featureNames.map((name) => Number(row[name])));
  }

  /**
   * ラベル（Survived）を抽出する
   */
  private extractLabels(data: DataFrame): number[] {
    const rows = data.toArray();
    return rows.map((row: any) => Number(row.Survived));
  }

  /**
   * データ前処理パイプライン
   */
  private preprocessData(df: IDataFrame): IDataFrame {
    let processed = this.preprocessAge(df);
    processed = this.encodeCategorical(processed);
    return processed;
  }

  /**
   * Age の欠損値を Pclass と Survived のグループ別中央値で補完
   */
  private preprocessAge(df: IDataFrame): IDataFrame {
    // グループ別の中央値マッピング
    const ageMapping: AgeMapping = {
      '1_0': 43,
      '1_1': 35,
      '2_0': 33,
      '2_1': 25,
      '3_0': 26,
      '3_1': 20,
    };

    // 欠損値を補完
    const processedData = df.select((row) => {
      const age = row.Age;
      if (age === null || age === undefined || age === '' || isNaN(Number(age))) {
        const key = `${row.Pclass}_${row.Survived}`;
        const medianAge = ageMapping[key];
        return { ...row, Age: medianAge };
      }
      return { ...row, Age: Number(age) };
    });

    return processedData;
  }

  /**
   * Sex をダミー変数に変換
   * drop_first=true により、male 列のみ作成（female は 0/1 で表現）
   * これにより多重共線性を回避
   */
  private encodeCategorical(df: IDataFrame): IDataFrame {
    const encoded = df.select((row) => {
      const male = row.Sex === 'male' ? 1 : 0;
      // Sex 列を除外して male 列を追加
      const result: any = {};
      Object.keys(row).forEach((key) => {
        if (key !== 'Sex') {
          result[key] = row[key];
        }
      });
      result.male = male;
      return result;
    });

    return encoded;
  }

  /**
   * モデルを訓練する
   */
  async train(): Promise<void>;
  // eslint-disable-next-line no-unused-vars
  train(x: number[][], y: number[]): void;
  async train(X?: number[][], y?: number[]): Promise<void> {
    // オーバーロードの実装
    if (X && y) {
      // 直接データを渡すケース
      this.model = new DecisionTreeClassifier({
        maxDepth: this.maxDepth,
        minNumSamples: 3,
      });

      this.model.train(X, y);
    } else {
      // loadData と splitData を使用するケース
      if (!this.X_train || !this.y_train) {
        throw new Error('Data not split. Call splitData() first.');
      }

      this.model = new DecisionTreeClassifier({
        maxDepth: this.maxDepth,
        minNumSamples: 3,
      });

      this.model.train(this.X_train, this.y_train);
    }
  }

  /**
   * 予測を実行する
   */
  predict(X: number[][]): number[] {
    if (this.model === null) {
      throw new Error('Model has not been trained yet. Call train() first.');
    }

    return this.model.predict(X);
  }

  /**
   * モデルを評価する（正解率を返す）
   */
  evaluate(): number {
    if (!this.X_test || !this.y_test || !this.model) {
      throw new Error('Model not trained or test data not available.');
    }

    const predictions = this.predict(this.X_test);
    const correct = predictions.filter((pred, i) => pred === this.y_test![i]).length;
    return correct / this.y_test.length;
  }

  /**
   * 訓練済みモデルをファイルに保存する
   */
  async save(filePath: string): Promise<void> {
    if (this.model === null) {
      throw new Error('Model has not been trained yet. Call train() first.');
    }

    const modelData = {
      maxDepth: this.maxDepth,
      model: this.model.toJSON(),
      featureNames: this.featureNames,
    };

    await fs.promises.writeFile(filePath, JSON.stringify(modelData, null, 2));
  }

  /**
   * 保存されたモデルをファイルから読み込む
   */
  async load(filePath: string): Promise<void> {
    try {
      await fs.promises.access(filePath);
    } catch {
      throw new Error(`Model file not found: ${filePath}`);
    }

    const fileContent = await fs.promises.readFile(filePath, 'utf-8');
    const modelData = JSON.parse(fileContent);

    this.model = DecisionTreeClassifier.load(modelData.model);
    this.featureNames = modelData.featureNames;
  }

  /**
   * データサイズを取得
   */
  getDataSize(): number {
    return this.data?.count() || 0;
  }

  /**
   * 訓練データサイズを取得
   */
  getTrainSize(): number {
    return this.X_train?.length || 0;
  }

  /**
   * テストデータサイズを取得
   */
  getTestSize(): number {
    return this.X_test?.length || 0;
  }
}
